### PR TITLE
Shashank/seq id flash attn

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
This PR does three flash attention-related things:
1. Adds support for sequence id masking when using flash attention 2.1.2 or higher. 
2. Gets rid of kv tensor repetition for grouped query attention when using flash attention 2.0.0 or higher.
3. Adds support for sliding window attention when using flash attention 2.3.0 or higher.

WandB link to the experiments: https://wandb.ai/mosaic-ml/seq_id_FA_final_tests

Loss and throughput curves for 125M model trained to chinchilla steps:
1. control is the main llmfoundry branch
2. treat is the Shashank/seq_id_flash_attn (the branch corresponding to this PR) with no config changes.
3. treat-seq-id-masking is the Shashank/seq_id_flash_attn (the branch corresponding to this PR) with sequence id masking turned on.
4. treat-sliding-window-100 is the Shashank/seq_id_flash_attn (the branch corresponding to this PR) with sliding window of size 100.
Screenshot 2023-12-01 at 9 37 53 AM

Screenshot 2023-12-01 at 9 38 48 AM

Screenshot 2023-12-01 at 9 43 01 AM

Screenshot 2023-12-01 at 9 43 44 AM

### CODE REVIEW

File: llmfoundry/models/layers/attention.py [hunk 1]
```diff
@@ -90,8 +90,12 @@ def scaled_multihead_dot_product_attention(
     training: bool = False,
     needs_weights: bool = False,
     multiquery: bool = False,
+    query_attention_mask_in_length: Optional[torch.Tensor] = None,
+    key_attention_mask_in_length: Optional[torch.Tensor] = None,
+    should_repeat_kv_for_gqa: Optional[bool] = True,
 ) -> tuple[torch.Tensor, Optional[torch.Tensor], Optional[tuple[torch.Tensor,
                                                                 torch.Tensor]]]:
+    del query_attention_mask_in_length, key_attention_mask_in_length, should_repeat_kv_for_gqa
```
Comment (vchiley) [inline llmfoundry/models/layers/attention.py:98]: @dakinggg do you think this should be `kwargs` then `del kwargs`
or should the interface for the 3 implementation stay EXACTLY the same.

File: llmfoundry/models/layers/attention.py [hunk 2]
```diff
@@ -640,6 +675,9 @@ def forward(
             dropout_p=self.attn_dropout_p,
             training=self.training,
             needs_weights=needs_weights,
+            query_attention_mask_in_length=query_attention_mask_in_length,
+            key_attention_mask_in_length=key_attention_mask_in_length,
+            should_repeat_kv_for_gqa= not is_flash_v2_installed(),
```
Comment (vchiley) [inline llmfoundry/models/layers/attention.py:680]:
```suggestion
            should_repeat_kv_for_gqa=not is_flash_v2_installed(),
```

did lint not complain about this?
looks weird to me, but yeah do what the linter says.

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 3]
```diff
@@ -222,9 +222,9 @@ def _validate_config(self) -> None:
             raise NotImplementedError(
                 'alibi only implemented with torch and triton attention.')
         if self.attn_config['attn_uses_sequence_id'] and self.attn_config[
-                'attn_impl'] not in ['torch', 'triton']:
+                'attn_impl'] not in ['torch', 'triton'] and not (self.attn_config['attn_impl']=='flash' and is_flash_v2_installed(v2_version='v2.1.2')):
             raise NotImplementedError(
-                'attn_uses_sequence_id only implemented with torch and triton attention.'
+                'attn_uses_sequence_id only implemented with torch, triton, and flash (v2.1.2 or higher) attention.'
```
Comment (vchiley) [inline llmfoundry/models/mpt/configuration_mpt.py:227]: i think seq id masking should work with fa 1 as well 🤔 
why is v2.1.2 specifically required? at the very least should it be `>=2.0.0`?

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 4]
```diff
@@ -222,9 +222,9 @@ def _validate_config(self) -> None:
             raise NotImplementedError(
                 'alibi only implemented with torch and triton attention.')
         if self.attn_config['attn_uses_sequence_id'] and self.attn_config[
-                'attn_impl'] not in ['torch', 'triton']:
+                'attn_impl'] not in ['torch', 'triton'] and not (self.attn_config['attn_impl']=='flash' and is_flash_v2_installed(v2_version='v2.1.2')):
```
Comment (vchiley) [inline llmfoundry/models/mpt/configuration_mpt.py:225]:
```suggestion
                'attn_impl']=='flash' and not is_flash_v2_installed(v2_version='v2.1.2')):
```

the current gating logic is pretty confusing.

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 5]
```diff
@@ -131,6 +131,44 @@ def gen_rotary_embedding(rope_head_dim: int, rope_impl: str, rope_theta: int,
...
+        if (sequence_id is not None) and attn_uses_sequence_id and (attn_impl == 'flash'):
+            query_attention_mask_in_length = torch.nn.functional.one_hot(sequence_id[:, -S:], num_classes=S).sum(dim=1)
+            # We use S as the number of classes while creating key_attention_mask_in_length instead of sequence_id.shape[-1] 
+            # because in case of inference, sequence_id.shape[-1] can become very large. In that case, the one_hot vectors 
+            # would've become very large as well.
+            key_attention_mask_in_length = torch.nn.functional.one_hot(sequence_id, num_classes=S).sum(dim=1)
+            # Since Flash Attention expects the masks to have same shape as the keys, we pad it with zeros.
+            key_attention_mask_in_length = torch.nn.functional.pad(key_attention_mask_in_length, (0, sequence_id.shape[-1] - S), value=0)
+                
+        return query_attention_mask_in_length,key_attention_mask_in_length
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:147]:
```suggestion
        return query_attention_mask_in_length, key_attention_mask_in_length
```

did lint not complain about that?
(again) looks weird to me, but yeah do what the linter says.

**Review by vchiley — COMMENTED** [review]
high level looks ok

in the pr description can you include figure showing mfu diff with and without masking and also figure showing convergence diff with and without masking

File: llmfoundry/models/layers/attention.py (hunk 2 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/attention.py:680]: I manually turned off lint while committing. Sorry about that :D . Will run it

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 3 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:227]: The function we use, [bert_padding.unpad_input_for_concatenated_sequences](https://github.com/Dao-AILab/flash-attention/blob/017716451d446e464dde9aca3a3c1ed2209caaa9/flash_attn/bert_padding.py#L125) was added in this [commit](https://github.com/Dao-AILab/flash-attention/commit/8f6f48d8a8c5e6c707ed3e9a2079dc1b12611122), which was added in v2.1.2.

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 4 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:225]: The reason I kept it that way was so that if in the future we add another type of attention which does not support sequence_id masking, then this will fail. If it does support, then it can simply be added to the list ['torch', 'triton']. Maybe I overthought lol.

**Review by dakinggg — COMMENTED** [review]
would be good to train some models to show equivalence of seq id with flash and other attention implementations

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 3 above)
Comment (vchiley) [reply llmfoundry/models/mpt/configuration_mpt.py:230]: Should it be `>=2.1.2` then?

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 6]
```diff
@@ -132,6 +132,74 @@ def gen_rotary_embedding(rope_head_dim: int, rope_impl: str, rope_theta: int,
...
+        1. Training with left padding is not supported in MPT (see https://github.com/mosaicml/llm-foundry/blob/1eecd4cb8e734499f77f6a35f657b8b20c0adfcb/llmfoundry/models/mpt/modeling_mpt.py#L407).
+        2. For generation with left padding, we only have a single sequence id per sample, so we don't need sequence id based sparse attention.
+
+    Args:
+        sequence_id (Union[None, torch.Tensor]): Tensor containing the sequence id for each token. Shape (batch_size, seq_len).
+        S (int): Sequence length
+        attn_uses_sequence_id (bool): Whether the attention uses sequence id based masking.
+        attn_impl (str): Attention implementation. This function is only creates attention_mask_in_length for flash attention.
+        attention_mask (Union[torch.Tensor, None]): Attention mask tensor of shape (batch_size, seq_len)
+    """
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:151]: given you have a good docstring, might as well add a `Return`

File: tests/test_flash_attn.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+                                                is_flash_v2_installed,
+                                                triton_flash_attn_fn)
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('kv_n_heads', [1, 4, 8])
+def test_gqa_kv_repetition(kv_n_heads: int):
+    # Test that flash attention v2 with GQA (kv_n_heads < n_heads) works the same
+    # whether we repeat the kv_n_heads explicitly or flash attention v2 handles it on its own.
+    if not is_flash_v2_installed():
```
Comment (dakinggg) [inline tests/test_flash_attn.py:19]: move this to `@pytest.skipif(<condition>)`

File: tests/test_flash_attn.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    assert torch.allclose(output_1, output_2)
+    assert torch.allclose(query_1.grad, query_2.grad)  # type: ignore
+    assert torch.allclose(key_1.grad, key_2.grad)  # type: ignore
+    assert torch.allclose(value_1.grad, value_2.grad)  # type: ignore
+
+
+@pytest.mark.gpu
+def test_seq_id_masking_FA_v2():
+    # Test that flash attention v2 with sequence id masking works correctly.
+    if not is_flash_v2_installed(v2_version='v2.1.2'):
```
Comment (dakinggg) [inline tests/test_flash_attn.py:87]: same

File: tests/test_flash_attn.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        assert torch.allclose(
+            value_1.grad[:, seq_range[0]:seq_range[1], :],  # type: ignore
+            value_2.grad)  # type: ignore
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('sliding_window_size', [1, 4, 8])
+def test_sliding_window(sliding_window_size: int):
+    # Test that sliding window attention works as expected.
+    if not is_flash_v2_installed('v2.3.0'):
```
Comment (dakinggg) [inline tests/test_flash_attn.py:173]: same

File: tests/test_model.py [hunk 10]
```diff
@@ -555,6 +555,119 @@ def test_mpt_creation(norm_type: str, no_bias: bool, tie_word_embeddings: bool):
...
+    'alibi': False,
+    'rope': True,
+    'rope_theta': 10000,
+    'rope_impl': 'hf',
+    'rope_hf_config': {
+        'type': 'no_scaling',
+        'factor': 1.0,
+    },
+}])
+@pytest.mark.parametrize('tie_word_embeddings', [True, False])
```
Comment (dakinggg) [inline tests/test_model.py:583]: do we need to ablate on tied word embeddings here? i dont think it really interacts with the code you're adding at all

**Review by dakinggg — APPROVED** [review]
LGTM, lets look into that slow test a bit before merging.

File: tests/test_flash_attn.py (hunk 7 above)
Comment (ShashankMosaicML) [reply tests/test_flash_attn.py]: Done.

File: tests/test_flash_attn.py (hunk 8 above)
Comment (ShashankMosaicML) [reply tests/test_flash_attn.py]: done

File: tests/test_flash_attn.py (hunk 9 above)
Comment (ShashankMosaicML) [reply tests/test_flash_attn.py]: done.

File: tests/test_model.py (hunk 10 above)
Comment (ShashankMosaicML) [reply tests/test_model.py]: done.
